mllib-local/src/test/scala/org/apache/spark/ml/linalg/BLASBenchmark.scala

```diff
@@ -0,0 +1,502 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.linalg
+
+import com.github.fommil.netlib.{BLAS => NetlibBLAS, F2jBLAS}
+
+import org.apache.spark.benchmark.{Benchmark, BenchmarkBase}
+
+/**
+ * Serialization benchmark for BLAS.
+ * To run this benchmark:
+ * {{{
+ * 1. without sbt: bin/spark-submit --class <this class> <spark mllib test jar>
+ * 2. build/sbt "mllib/test:runMain <this class>"
+ * 3. generate result: SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "mllib/test:runMain <this class>"
+ *    Results will be written to "benchmarks/BLASBenchmark-results.txt".
+ * }}}
+ */
+object BLASBenchmark extends BenchmarkBase {
+
+  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
+
+    val iters = 1e2.toInt
+    val rnd = new scala.util.Random(0)
+
+    val f2jBLAS = new F2jBLAS
+    val nativeBLAS = NetlibBLAS.getInstance
+    val vectorBLAS =
+      try {
+        // scalastyle:off classforname
+        Class.forName("org.apache.spark.ml.linalg.VectorizedBLAS", true,
+                        Option(Thread.currentThread().getContextClassLoader)
+                          .getOrElse(getClass.getClassLoader))
+             .newInstance()
+             .asInstanceOf[NetlibBLAS]
+        // scalastyle:on classforname
+      } catch {
+        case _: Throwable => new F2jBLAS
+      }
```

Comment:
Hmm, for this benchmark, do we want to fallback to `F2jBLAS`? I think we want to benchmark `VectorizedBLAS` actually.

I saw this benchmark doesn't add cases of native BLAS and vectorized BLAS if they aren't available, but if so, is it useful to run the benchmark?

Reply:
It's indeed not necessarily helpful to run the benchmark if there is no `VectorizedBLAS` nor `NetblibRefBLAS`/`NetlibSystemBLAS`. However, AFAICT, this benchmark is only run manually and we can then assume that whoever is running it is trying to compare two implementations. If you feel strongly about not running the benchmark if only `F2jBLAS` implementation is available, let me know and I'll modify it to do so.